### ISSUE TITLE
Copy hosting details to clipboard

### ISSUE THREAD

**Issue by thecoshman** [thecoshman/http#10]
When the server starts up, it should also copy the details to the clipboard, this should be the IP for the machine, not 'localhost' so you can paste it into some chat easier

**Comment by nabijaczleweli** [comment]
Retracted in https://github.com/thecoshman/http/pull/14/commits/d4755135cc29e2686d9a61881bf72b5b4677794b

**Comment by nabijaczleweli** [comment]
So what should we do with this? [On Linux it requires `xorg-dev`](https://github.com/aweinstock314/rust-clipboard#prerequisites) and you had some trouble with that?

**Comment by thecoshman** [comment]
It's not xorg-dev that I had problems with, I think it was simply that my version of rustc was out of date. Maybe try re-adding these changes... possibly to a side branch for now... if I it works for me, we can merge them to dev, if not, we can just leave it there until we work out why it's not 'plug and play', rather than reverting back and forth

**Comment by nabijaczleweli** [comment]
@thecoshman check out `feature/10-clipboard-external-ip` branch to see if it compiles so we can put it into #32

**Comment by thecoshman** [comment]
Changed my mind on this one, copying into the clipboard automatically is messy for the user. Having a cli arg to say you want to do this is stupid.

Working out _what_ to copy is also very tedious.

**Comment by nabijaczleweli** [comment]
Discussion:

> [14:09] Pirate: oh my, just noticed issue #10 @((наб *)(войд)())() I should get around to testing that at some stage :S
> [14:10] ((наб *)(войд)())(): you shouldn't, it's a stupid idea
> [14:10] Pirate: I was just thinking about it really...
> [14:11] Pirate: You print that hosting details at startup, right?
> [14:11] Pirate: making them trivial to copy from the console
> [14:11] ((наб *)(войд)())(): no
> [14:11] ((наб *)(войд)())(): i print information
> [14:12] ((наб *)(войд)())(): each host has at least two network interfaces
> [14:12] ((наб *)(войд)())(): what's the output for 
> ```
> nabijaczleweli@tarta:~$ networkctl
> IDX LINK             TYPE               OPERATIONAL SETUP
>   1 lo               loopback           carrier     unmanaged
>   2 eno1             ether              carrier     configuring
>   3 eno2             ether              carrier     configuring
>   4 eno3             ether              carrier     configuring
>   5 eno4             ether              carrier     configuring
>   6 bond1            bond               routable    configured
> 
> 6 links listed.
> ```
> [14:13] ((наб *)(войд)())(): or
> ```
> nabijaczleweli@nabtop:~$ networkctl
> IDX LINK             TYPE               OPERATIONAL SETUP
>   1 lo               loopback           carrier     unmanaged
>   2 lan1             ether              routable    configured
>   3 wifi1            wlan               routable    configured
> 
> 3 links listed.
> ```
> [14:14] ((наб *)(войд)())(): do we do rdns or not and copy with hostname/FQDN/IP
> [14:14] Pirate: when you start up http what do you print?
> [14:14] ((наб *)(войд)())(): read the manpage for the variations
> [14:14] Pirate: :open_mouth:
> [14:16] Pirate: by default, localhost:<port> or 0.0.0.0:<port> would work, right? Or you used the cli arg to bind a certain address...
> [14:17] ((наб *)(войд)())(): 0.0.0.0 won't
> [14:17] ((наб *)(войд)())(): localhost might not, 127.0.0.1 is useless anyway since if you're pasting it you want to paste it into some chat probably
> [14:17] Pirate: :thinking:
> [14:18] Pirate: Ergh... depends what you want to use it for...
> [14:18] Pirate: for testing it yourself, lcoalhost is obviously fine, but useless for others
> [14:20] Pirate: oh ok, so I'm kinda thinking that copying the details to clipboard is a none nice idea... it's hard to get it work on all platforms, and is going to be a pain if you don't want it copied
> [14:21] Pirate: Better to just try put the details on console for someone to copy from there
> [14:21] ((наб *)(войд)())(): the one that we know about (the port) is in the console
> [14:22] Pirate: if -a  is used, we know the IP for sure, so having <ip>:<port> that you can highlight to copy would be nice
> [14:23] ((наб *)(войд)())(): loopback might also work, NAT may be in place
> [14:23] Pirate: if a computer has multiple network devices, could we list all them IPs?
> [14:23] ((наб *)(войд)())(): that's hilarious
> [14:23] Pirate: if NAT is in place, it's not the concern of http
> [14:24] Pirate: that's configured externally, so someone else has to deal with that
> [14:24] ((наб *)(войд)())(): we could but it's very system-dependent
> [14:24] Pirate: listing the IPs of network devices?
> [14:24] ((наб *)(войд)())(): plus, some/most/all ifaces will have any combination of v4 and v6 addresses
> [14:27] Pirate: yeah... could get very messy
> [14:27] ((наб *)(войд)())(): this is too much work for a thing too broken for anyone to ever actually use
> [14:28] Pirate: closed :smile:


Should the `feature/10-clipboard-external-ip` branch get deleted:

```
$ git log master..feature/10-clipboard-external-ip
commit 854887459814ade5ea5c936414f99294a110c913 (origin/feature/10-clipboard-external-ip, feature/10-clipboard-external-ip)
gpg: Signature made 01/12/17 14:35:12 Central European Standard Time
gpg:                using RSA key BCFD0B018D2658F1
gpg: Good signature from "nabijaczleweli <nabijaczleweli@gmail.com>" [ultimate]
gpg:                 aka "keybase.io/nabijaczleweli <nabijaczleweli@keybase.io>" [ultimate]
Author: nabijaczleweli <nabijaczleweli@gmail.com>
Date:   Thu Jan 12 14:20:51 2017 +0100

    Make clipboard support optional. Don't compile with clipboard on Travis

    Use --no-default-features to build w/o clipboard

    Add notice to manpage

commit 7472a6a5790162137b5da9c5bb8c529557475e36
gpg: Signature made 01/12/17 02:22:25 Central European Standard Time
gpg:                using RSA key BCFD0B018D2658F1
gpg: Good signature from "nabijaczleweli <nabijaczleweli@gmail.com>" [ultimate]
gpg:                 aka "keybase.io/nabijaczleweli <nabijaczleweli@keybase.io>" [ultimate]
Author: nabijaczleweli <nabijaczleweli@gmail.com>
Date:   Sat Dec 31 00:20:07 2016 +0100

    Revert "Revert "Copy external URL to the clipboard on server start" and "Install xorg-dev to build rust-clipboard properly""

    @thecoshman check this again

    This reverts commit d4755135cc29e2686d9a61881bf72b5b4677794b,
    re-adding 7526d383415c835a47a53026efb09bbc1bc2ecbd, and
    b77280782a983e56baf0626b3ef08d95f94e6bfd.

    Closes #10
```

```
$ git format-patch master..feature/10-clipboard-external-ip
0001-Revert-Revert-Copy-external-URL-to-the-clipboard-on-.patch
0002-Make-clipboard-support-optional.-Don-t-compile-with-.patch
```

```patch
$ cat 0001-Revert-Revert-Copy-external-URL-to-the-clipboard-on-.patch
From 7472a6a5790162137b5da9c5bb8c529557475e36 Mon Sep 17 00:00:00 2001
From: nabijaczleweli <nabijaczleweli@gmail.com>
Date: Sat, 31 Dec 2016 00:20:07 +0100
Subject: [PATCH 1/2] Revert "Revert "Copy external URL to the clipboard on
 server start" and "Install xorg-dev to build rust-clipboard properly""

@thecoshman check this again

This reverts commit d4755135cc29e2686d9a61881bf72b5b4677794b,
re-adding 7526d383415c835a47a53026efb09bbc1bc2ecbd, and
b77280782a983e56baf0626b3ef08d95f94e6bfd.

Closes #10
---
 .travis.yml     | 11 ++++++++++-
 Cargo.toml      |  7 ++++++-
 src/main.rs     | 10 ++++++++++
 src/util/mod.rs | 12 +++++++++++-
 4 files changed, 37 insertions(+), 3 deletions(-)

diff --git a/.travis.yml b/.travis.yml
index 8c94ade..1452abb 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -1,6 +1,15 @@
 sudo: false
 language: generic
-cache: cargo
+cache:
+  apt: true
+  cargo: true
+
+addons:
+  apt:
+    packages:
+      - xorg-dev
+    sources:
+      - r-packages-precise
 
 matrix:
   include:
diff --git a/Cargo.toml b/Cargo.toml
index 0bc8f4d..7663ace 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -13,17 +13,22 @@ authors = ["thecoshman <thecoshman@gmail.com>",
 [dependencies]
 lazy_static = "0.2"
 mime_guess = "1.8"
+clipboard = "0.1"
 lazysort = "0.1"
 brotli2 = "0.2"
 base64 = "0.2"
 flate2 = "0.2"
+hyper = "0.9"
 bzip2 = "0.3"
 clap = "2.20"
-iron = "0.4"
 time = "0.1"
 url = "1.2"
 md6 = "1.1"
 
+[dependencies.iron]
+version = "0.4"
+features = ["ssl"]
+
 
 [[bin]]
 name = "http"
diff --git a/src/main.rs b/src/main.rs
index e5fecb5..1991d7c 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -1,11 +1,13 @@
 #[macro_use]
 extern crate lazy_static;
 extern crate mime_guess;
+extern crate clipboard;
 extern crate lazysort;
 extern crate brotli2;
 extern crate base64;
 extern crate flate2;
 extern crate bzip2;
+extern crate hyper;
 #[macro_use]
 extern crate clap;
 extern crate iron;
@@ -25,6 +27,7 @@ pub use options::Options;
 use iron::Iron;
 use std::io::stderr;
 use std::process::exit;
+use clipboard::ClipboardContext;
 
 
 fn main() {
@@ -59,6 +62,13 @@ fn result_main() -> Result<(), Error> {
     });
 
     println!("Hosting \"{}\" on port {}...", opts.hosted_directory.0, responder.socket.port());
+    if let Some(self_ip) = util::response_body("https://api.ipify.org") {
+        if let Ok(mut clpbrd) = ClipboardContext::new() {
+            if clpbrd.set_contents(format!("{}:{}", self_ip, responder.socket.port())).is_ok() {
+                println!("Externally-accessible URL is in the clipboard.");
+            }
+        }
+    }
     println!("Ctrl-C to stop.");
     println!();
 
diff --git a/src/util/mod.rs b/src/util/mod.rs
index 7300dbc..0ce7eee 100644
--- a/src/util/mod.rs
+++ b/src/util/mod.rs
@@ -8,12 +8,13 @@ use std::f64;
 use std::cmp;
 use iron::Url;
 use std::fs::File;
+use hyper::Client;
 use std::path::Path;
 use std::borrow::Cow;
 use url::percent_encoding;
 use std::collections::HashMap;
 use time::{self, Duration, Tm};
-use std::io::{BufReader, BufRead};
+use std::io::{BufReader, BufRead, Read};
 use mime_guess::get_mime_type_str;
 
 pub use self::content_encoding::*;
@@ -124,6 +125,15 @@ pub fn html_response<S: AsRef<str>>(data: &str, format_strings: &[S]) -> String
                        |d, (k, v)| d.replace(&format!("{{{}}}", k), v))
 }
 
+/// Get the response body from the provided URL.
+pub fn response_body(url: &str) -> Option<String> {
+    Client::new().get(url).send().ok().map(|mut r| {
+        let mut body = String::new();
+        r.read_to_string(&mut body).unwrap();
+        body
+    })
+}
+
 /// Return the path part of the URL.
 ///
 /// # Example
-- 
2.25.1

```

```patch
$ cat 0002-Make-clipboard-support-optional.-Don-t-compile-with-.patch
From 854887459814ade5ea5c936414f99294a110c913 Mon Sep 17 00:00:00 2001
From: nabijaczleweli <nabijaczleweli@gmail.com>
Date: Thu, 12 Jan 2017 14:20:51 +0100
Subject: [PATCH 2/2] Make clipboard support optional. Don't compile with
 clipboard on Travis

Use --no-default-features to build w/o clipboard

Add notice to manpage
---
 .travis.yml | 15 ++++-----------
 Cargo.toml  |  8 +++++++-
 http.md     |  6 ++++++
 src/main.rs | 13 +++++++++----
 4 files changed, 26 insertions(+), 16 deletions(-)

diff --git a/.travis.yml b/.travis.yml
index 1452abb..21925f5 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -4,13 +4,6 @@ cache:
   apt: true
   cargo: true
 
-addons:
-  apt:
-    packages:
-      - xorg-dev
-    sources:
-      - r-packages-precise
-
 matrix:
   include:
     - env: LANGUAGE=Rust DEPLOY=true DEPLOY_FILE="$TRAVIS_BUILD_DIR/../http-$TRAVIS_TAG"
@@ -34,13 +27,13 @@ install:
   - if [ "$LANGUAGE" == "Ruby" ]; then gem install ronn; fi
 
 script:
-  - if [ "$LANGUAGE" == "Rust" ]; then cargo build --verbose; fi
-  - if [ "$LANGUAGE" == "Rust" ]; then cargo test  --verbose; fi
+  - if [ "$LANGUAGE" == "Rust" ]; then cargo build --no-default-features --verbose; fi
+  - if [ "$LANGUAGE" == "Rust" ]; then cargo test  --no-default-features --verbose; fi
   -
-  - if [ "$LANGUAGE" == "Rust" ] && [ "$DEPLOY" ] && [ "$TRAVIS_TAG" ]; then cargo build --verbose --release; fi
+  - if [ "$LANGUAGE" == "Rust" ] && [ "$DEPLOY" ] && [ "$TRAVIS_TAG" ]; then cargo build --no-default-features --verbose --release; fi
   - if [ "$LANGUAGE" == "Rust" ] && [ "$CLIPPY" ]; then
       cargo install -f clippy;
-      cargo clippy;
+      cargo clippy --no-default-features;
     fi
   -
   - if [ "$LANGUAGE" == "Ruby" ]; then
diff --git a/Cargo.toml b/Cargo.toml
index 7663ace..5107036 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -13,7 +13,6 @@ authors = ["thecoshman <thecoshman@gmail.com>",
 [dependencies]
 lazy_static = "0.2"
 mime_guess = "1.8"
-clipboard = "0.1"
 lazysort = "0.1"
 brotli2 = "0.2"
 base64 = "0.2"
@@ -29,6 +28,13 @@ md6 = "1.1"
 version = "0.4"
 features = ["ssl"]
 
+[dependencies.clipboard]
+version = "0.1"
+optional = true
+
+[features]
+default = ["clipboard"]
+
 
 [[bin]]
 name = "http"
diff --git a/http.md b/http.md
index 7a78414..098501e 100644
--- a/http.md
+++ b/http.md
@@ -15,6 +15,11 @@ be used via Linux CLI to quickly take the current directory and serve it over
 HTTP. Everything should have sensible defaults such that you do not *have* to
 pass parameters like what port to use.
 
+By default `http` is built with clipboard support -- it'll copy the
+externally-accessible URL to the clipboard, which requires the `xorg-dev`
+package on Linux. If, for some reason, that's inaccessible on your system,
+compile with `--no-default-features`.
+
 ## OPTIONS
 
   [DIR]
@@ -82,6 +87,7 @@ pass parameters like what port to use.
     Example output:
       p:\Rust\http> http
       Hosting "." on port 8000...
+      Externally-accessible URL is in the clipboard.
       Ctrl-C to stop.
 
       127.0.0.1:47880 was served directory listing for \\?\P:\Rust\http
diff --git a/src/main.rs b/src/main.rs
index 1991d7c..cbb4aad 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -1,6 +1,7 @@
 #[macro_use]
 extern crate lazy_static;
 extern crate mime_guess;
+#[cfg(feature = "clipboard")]
 extern crate clipboard;
 extern crate lazysort;
 extern crate brotli2;
@@ -27,6 +28,7 @@ pub use options::Options;
 use iron::Iron;
 use std::io::stderr;
 use std::process::exit;
+#[cfg(feature = "clipboard")]
 use clipboard::ClipboardContext;
 
 
@@ -62,10 +64,13 @@ fn result_main() -> Result<(), Error> {
     });
 
     println!("Hosting \"{}\" on port {}...", opts.hosted_directory.0, responder.socket.port());
-    if let Some(self_ip) = util::response_body("https://api.ipify.org") {
-        if let Ok(mut clpbrd) = ClipboardContext::new() {
-            if clpbrd.set_contents(format!("{}:{}", self_ip, responder.socket.port())).is_ok() {
-                println!("Externally-accessible URL is in the clipboard.");
+    #[cfg(feature = "clipboard")]
+    {
+        if let Some(self_ip) = util::response_body("https://api.ipify.org") {
+            if let Ok(mut clpbrd) = ClipboardContext::new() {
+                if clpbrd.set_contents(format!("{}:{}", self_ip, responder.socket.port())).is_ok() {
+                    println!("Externally-accessible URL is in the clipboard.");
+                }
             }
         }
     }
-- 
2.25.1

```

**Comment by thecoshman** [comment]
I don't really see any harm in keep the branch around as like a back up... who knows, maybe it'll be handy to see how to do some stuff for something else in future. I don't mind either way really.
